File: packages/paperdoll/src/direction.ts
// The 8 facing rows every sprite sheet is authored in. Order is fixed and
// load-bearing: sheet row N *is* Direction N, and the layer-order matrix indexes
// by it. See specs/paperdoll for the convention.
//
// Index 0 is South (facing the camera in a top-down/2d view); indices advance
// clockwise-from-south through the diagonals. The engine's yaw is 0 when an
// entity faces +Z (screen-down / south) and increases turning toward +X (east),
// so direction index runs *opposite* yaw — captured once, here, and nowhere
// else in the codebase.

export const DIRECTION_COUNT = 8;

export type Direction = 0 | 1 | 2 | 3 | 4 | 5 | 6 | 7;

/** S, SW, W, NW, N, NE, E, SE — human labels for the row order, for debug/tests. */
export const DIRECTION_LABELS = ["S", "SW", "W", "NW", "N", "NE", "E", "SE"] as const;

const SECTOR = (Math.PI * 2) / DIRECTION_COUNT; // 45°

/**
 * Quantize an engine yaw (radians) to a sheet row. Rounds to the nearest of 8
 * sectors, so a yaw exactly between two rows snaps to the lower-numbered one.
 */
export function directionFromYaw(yaw: number): Direction {
  // -yaw because direction runs opposite yaw (see module doc). +DIRECTION_COUNT
  // before the final mod so negative yaws don't produce a negative index.
  const idx = ((Math.round(-yaw / SECTOR) % DIRECTION_COUNT) + DIRECTION_COUNT) % DIRECTION_COUNT;
  return idx as Direction;
}

/**
 * Billboard variant: the row shown depends on where the camera stands, not just
 * where the entity faces. Passing the camera's azimuth (its heading around the
 * world Y axis) yields the direction the entity presents *to that camera*, so
 * orbiting a standing entity walks through all 8 rows. `cameraAzimuth === 0`
 * collapses to `directionFromYaw`.
 */
export function directionFromYawRelativeToCamera(yaw: number, cameraAzimuth: number): Direction {
  return directionFromYaw(yaw - cameraAzimuth);
}

File: crates/animation/src/direction.rs
//! Yaw → 8-way sheet row. Mirrors the (now-removed) TypeScript quantizer so the
//! engine and the sprite agree on which row is shown — the boundary sends the
//! resolved direction, TS never recomputes it.
//!
//! Index 0 is South (facing the camera in a top-down view); indices advance
//! clockwise-from-south through the diagonals. Engine yaw is 0 facing +Z (south)
//! and increases toward +X (east), so the row index runs opposite yaw.

use std::f32::consts::TAU;

pub const DIRECTION_COUNT: u8 = 8;

/// Quantize an engine yaw (radians) to a direction row 0..=7.
pub fn direction_from_yaw(yaw: f32) -> u8 {
    let sector = TAU / DIRECTION_COUNT as f32; // 45°
    // -yaw because the row runs opposite yaw; rem_euclid keeps it non-negative.
    let idx = (-yaw / sector).round() as i32;
    idx.rem_euclid(DIRECTION_COUNT as i32) as u8
}

/// Billboard variant: the row an entity presents to a camera at `camera_azimuth`
/// (its heading around world Y). Collapses to `direction_from_yaw` at azimuth 0.
pub fn direction_from_yaw_relative(yaw: f32, camera_azimuth: f32) -> u8 {
    direction_from_yaw(yaw - camera_azimuth)
}

#[cfg(test)]
mod tests {
    use super::*;
    use std::f32::consts::PI;

    #[test]
    fn cardinals_and_diagonals() {
        assert_eq!(direction_from_yaw(0.0), 0); // S
        assert_eq!(direction_from_yaw(-PI / 4.0), 1); // SW
        assert_eq!(direction_from_yaw(-PI / 2.0), 2); // W
        assert_eq!(direction_from_yaw(-3.0 * PI / 4.0), 3); // NW
        assert_eq!(direction_from_yaw(PI), 4); // N
        assert_eq!(direction_from_yaw(3.0 * PI / 4.0), 5); // NE
        assert_eq!(direction_from_yaw(PI / 2.0), 6); // E
        assert_eq!(direction_from_yaw(PI / 4.0), 7); // SE
    }

    #[test]
    fn wraps_negative_and_over_two_pi() {
        assert_eq!(direction_from_yaw(-PI), 4); // -π == π == N
        assert_eq!(direction_from_yaw(2.0 * PI), 0); // full turn == S
    }

    #[test]
    fn camera_relative_rotates_the_row() {
        assert_eq!(direction_from_yaw_relative(PI / 2.0, 0.0), direction_from_yaw(PI / 2.0));
        assert_eq!(direction_from_yaw_relative(0.0, PI / 2.0), 2); // faces S, camera a quarter-turn away → W
    }
}
